presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/ResourceGroup.java

```diff
@@ -0,0 +1,486 @@
+    // Sub groups with queued queries, that have capacity to run them
+    // That is, they must return true when internalStartNext() is called on them
+    @GuardedBy("root")
+    private final Queue<ResourceGroup> eligibleSubGroups = new LinkedHashQueue<>();
+    @GuardedBy("root")
+    private final Set<ResourceGroup> dirtySubGroups = new HashSet<>();
+    @GuardedBy("root")
+    private long maxMemoryBytes;
+    @GuardedBy("root")
+    private int maxRunningQueries;
+    @GuardedBy("root")
+    private int maxQueuedQueries;
+    @GuardedBy("root")
+    private int descendantRunningQueries;
+    @GuardedBy("root")
+    private int descendantQueuedQueries;
+    @GuardedBy("root")
+    private long cachedMemoryUsageBytes;
+    @GuardedBy("root")
+    private final Queue<QueryExecution> queuedQueries = new LinkedHashQueue<>();
+    @GuardedBy("root")
+    private final Set<QueryExecution> runningQueries = new HashSet<>();
+
+    protected ResourceGroup(Optional<ResourceGroup> parent, String name, int maxRunningQueries, int maxQueuedQueries,  long maxMemoryBytes, Executor executor)
+    {
+        this.parent = parent;
+        this.maxRunningQueries = maxRunningQueries;
+        this.maxQueuedQueries = maxQueuedQueries;
+        this.maxMemoryBytes = maxMemoryBytes;
+        this.executor = requireNonNull(executor, "executor is null");
+        requireNonNull(name, "name is null");
+        if (parent.isPresent()) {
+            id = new ResourceGroupId(parent.get().id, name);
+            root = parent.get().root;
+        }
+        else {
+            id = new ResourceGroupId(name);
+            root = this;
+        }
+    }
+
+    public void setMaxMemory(DataSize maxMemory)
+    {
+        synchronized (root) {
+            boolean oldCanRun = canRunMore();
+            this.maxMemoryBytes = maxMemory.toBytes();
+            if (canRunMore() != oldCanRun) {
+                updateEligiblility();
+            }
+        }
+    }
+
+    public void setMaxRunningQueries(int maxRunningQueries)
+    {
+        synchronized (root) {
+            boolean oldCanRun = canRunMore();
+            this.maxRunningQueries = maxRunningQueries;
+            if (canRunMore() != oldCanRun) {
+                updateEligiblility();
+            }
+        }
+    }
+
+    public void setMaxQueuedQueries(int maxQueuedQueries)
+    {
+        synchronized (root) {
+            this.maxQueuedQueries = maxQueuedQueries;
+        }
+    }
+
+    public ResourceGroup addSubGroup(String name, int maxRunningQueries, int maxQueuedQueries, DataSize maxMemory)
+    {
+        synchronized (root) {
+            checkArgument(runningQueries.isEmpty() && queuedQueries.isEmpty(), "Cannot add sub group to %s. It is running queries.", id);
+            ResourceGroup subGroup = new ResourceGroup(Optional.of(this), name, maxRunningQueries, maxQueuedQueries, maxMemory.toBytes(), executor);
+            subGroups.add(subGroup);
```

Comment:
`ResourceGroup` does not implement `equals` and `hashCode` so I think you are going to get duplicates